packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/CopyFromRealmTests.kt

```diff
@@ -0,0 +1,755 @@
+            .directory(tmpDir)
+            .build()
+        realm = Realm.open(configuration)
+    }
+
+    @AfterTest
+    fun tearDown() {
+        if (!realm.isClosed()) {
+            realm.close()
+        }
+        PlatformUtils.deleteTempDir(tmpDir)
+    }
+
+    @Test
+    fun primitiveValues() { // This also checks that any default values set in the class are being overridden correctly.
+        val type = Sample::class
+        val schemaProperties = type.realmObjectCompanionOrThrow().io_realm_kotlin_schema().properties
+        val fields: Map<String, KProperty1<*, *>> = type.realmObjectCompanionOrThrow().io_realm_kotlin_fields
+
+        // Dynamically set data on the Sample object
+        val originalObject = Sample()
+        schemaProperties.forEach { prop: RealmProperty ->
+            if (prop.type is ValuePropertyType) {
+                val accessor: KMutableProperty1<BaseRealmObject, Any?> = fields[prop.name] as KMutableProperty1<BaseRealmObject, Any?>
+                val fieldValue: Any? = createPrimitiveValueData(accessor)
+                accessor.set(originalObject, fieldValue)
+            }
+        }
+
+        // Round-trip object through `copyToRealm` and `copyFromRealm`.
+        val unmanagedCopy = realm.writeBlocking {
+            copyToRealm(originalObject).copyFromRealm()
+        }
+
+        // Close Realm to ensure data is decoupled from Realm
+        realm.close()
+
+        // Validate that all primitive list fields were round-tripped correctly.
+        schemaProperties.forEach { prop: RealmProperty ->
+            if (prop.type is ValuePropertyType) {
+                val accessor: KMutableProperty1<BaseRealmObject, Any?> = fields[prop.name] as KMutableProperty1<BaseRealmObject, Any?>
+                val value: Any? = createPrimitiveValueData(accessor)
+
+                if (prop.type.storageType == RealmStorageType.BINARY) {
+                    val copiedValue = accessor.get(unmanagedCopy) as ByteArray?
+                    assertContentEquals(value as ByteArray?, copiedValue, "${prop.name} failed")
+                } else {
+                    val copiedValue = accessor.get(unmanagedCopy) as Any?
+                    assertEquals(value, copiedValue, "${prop.name} failed")
+                }
+            }
+        }
+    }
+
+    @Test
+    fun realmObjectReferences() {
+        val innerSample = Sample().apply { stringField = "inner" }
+
+        val insertedObj = realm.writeBlocking {
+            copyToRealm(Sample().apply { nullableObject = innerSample })
+        }
+        val unmanagedObj: Sample = insertedObj.copyFromRealm()
+        assertNotSame(insertedObj, unmanagedObj)
+        assertNotNull(unmanagedObj.nullableObject)
+        val innerCopy = unmanagedObj.nullableObject!!
+        assertFalse(innerCopy.isManaged())
+        assertEquals("inner", innerCopy.stringField)
+    }
+
+    @Test
+    fun embeddedObjectReferences() {
```

Comment:
Maybe also a test of `embeddedObject.copyFromRealm()`, since you can actually copy it out without a parent. I can't see why it should work, but would be nice to have the test to catch regressions if changing the implementation.